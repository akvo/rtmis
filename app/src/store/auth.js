import { Store } from 'pullstate';
/**
 * Server config
 */
const AuthState = new Store({
  authenticationCode: '',
  useAuthenticationCode: false, // using code for authentication
  username: '',
  password: '',
  token: null,
});

export default AuthState;
