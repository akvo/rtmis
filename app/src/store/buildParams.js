import { Store } from 'pullstate';
import defaultBuildParams from '../build';

const BuildParamsState = new Store({
  authenticationType: defaultBuildParams?.authenticationType || [
    'code_assignment',
    'username',
    'password',
  ],
  serverURL: defaultBuildParams?.serverURL,
  debugMode: defaultBuildParams?.debugMode || false,
  dataSyncInterval: defaultBuildParams?.dataSyncInterval || 300,
  errorHandling: defaultBuildParams?.errorHandling || true,
  loggingLevel: defaultBuildParams?.loggingLevel || 'verbose',
  appVersion: defaultBuildParams?.appVersion || '1.0.0',
  gpsThreshold: defaultBuildParams?.gpsThreshold || 20,
});

export default BuildParamsState;
