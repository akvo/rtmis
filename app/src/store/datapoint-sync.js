import { Store } from 'pullstate';

const DatapointSyncState = new Store({
  inProgress: false,
  progress: 0,
  added: false,
  completed: false,
});

export default DatapointSyncState;
