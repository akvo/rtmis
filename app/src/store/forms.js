import { Store } from 'pullstate';

const FormState = new Store({
  form: {},
  currentValues: {}, // answers
  questionGroupListCurrentValues: {}, // answers for question group list component
  visitedQuestionGroup: [], // to store visited question group id
  surveyDuration: 0,
  surveyStart: null,
  cascades: {},
  lang: 'en',
});

export default FormState;
