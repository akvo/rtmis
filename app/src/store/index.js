export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
