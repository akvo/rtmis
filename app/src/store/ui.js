import { Store } from 'pullstate';

const UIState = new Store({
  isDarkMode: false, // if isDarkMode = false then Theme= light
  lang: 'en',
  fontSize: 16,
  currentPage: 'GetStarted',
  online: false,
  networkType: null,
  isManualSynced: false,
});

export default UIState;
