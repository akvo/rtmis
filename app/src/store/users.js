import { Store } from 'pullstate';

const UserState = new Store({
  id: null,
  name: '',
  password: '',
  token: null,
  syncWifiOnly: false,
  syncInterval: 300,
  forms: [],
});

export default UserState;
