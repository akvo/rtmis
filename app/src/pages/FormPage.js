import React, { useState, useEffect, useCallback, useMemo } from 'react';
import {
  Platform,
  ToastAndroid,
  BackHandler,
  ActivityIndicator,
  StyleSheet,
  View,
} from 'react-native';
import { Button, Dialog, Text } from '@rneui/themed';
import Icon from 'react-native-vector-icons/Ionicons';
import { FormContainer } from '../form';
import { SaveDialogMenu, SaveDropdownMenu } from '../form/support';
import { BaseLayout } from '../components';
import { crudDataPoints } from '../database/crud';
import { UserState, UIState, FormState } from '../store';
import { getDurationInMinutes } from '../form/lib';
import { i18n } from '../lib';

const FormPage = ({ navigation, route }) => {
  const selectedForm = FormState.useState((s) => s.form);
  const surveyDuration = FormState.useState((s) => s.surveyDuration);
  const surveyStart = FormState.useState((s) => s.surveyStart);
  const currentValues = FormState.useState((s) => s.currentValues);
  const userId = UserState.useState((s) => s.id);
  const [onSaveFormParams, setOnSaveFormParams] = useState({});
  const [showDialogMenu, setShowDialogMenu] = useState(false);
  const [showDropdownMenu, setShowDropdownMenu] = useState(false);
  const [showExitConfirmationDialog, setShowExitConfirmationDialog] = useState(false);
  const activeLang = UIState.useState((s) => s.lang);
  const trans = i18n.text(activeLang);

  const currentFormId = route?.params?.id;
  // continue saved submission
  const savedDataPointId = route?.params?.dataPointId;
  const isNewSubmission = route?.params?.newSubmission;
  const [currentDataPoint, setCurrentDataPoint] = useState({});
  const [loading, setLoading] = useState(false);

  const refreshForm = () => {
    FormState.update((s) => {
      s.currentValues = {};
      s.questionGroupListCurrentValues = {};
      s.visitedQuestionGroup = [];
      s.cascades = {};
      s.surveyDuration = 0;
    });
  };

  useEffect(() => {
    const backHandler = BackHandler.addEventListener('hardwareBackPress', () => {
      const values = onSaveFormParams?.values;
      if (values && Object.keys(values).length) {
        setShowDialogMenu(true);
        return true;
      }
      refreshForm();
      return false;
    });
    return () => backHandler.remove();
  }, [onSaveFormParams]);

  useEffect(() => {
    if (!isNewSubmission) {
      fetchSavedSubmission().catch((e) => console.error('[Fetch Data Point Failed]: ', e));
    }
  }, [isNewSubmission]);

  const fetchSavedSubmission = useCallback(async () => {
    setLoading(true);
    const dpValue = await crudDataPoints.selectDataPointById({ id: savedDataPointId });
    setCurrentDataPoint(dpValue);
    if (dpValue?.json && Object.keys(dpValue.json)?.length) {
      FormState.update((s) => {
        s.currentValues = dpValue.json;
        s.questionGroupListCurrentValues = dpValue.json;
      });
    }
    setLoading(false);
  }, [savedDataPointId]);

  const formJSON = useMemo(() => {
    if (!selectedForm?.json) {
      return {};
    }
    return JSON.parse(selectedForm.json);
  }, [selectedForm]);

  const onSaveCallback = useCallback((values) => {
    const state = { values };
    setOnSaveFormParams(state);
  }, []);

  const handleOnPressArrowBackButton = () => {
    const values = onSaveFormParams?.values;
    if (values && Object.keys(values).length) {
      setShowDialogMenu(true);
      return;
    }
    refreshForm();
    return navigation.goBack();
  };

  const handleOnSaveAndExit = async () => {
    const { values } = onSaveFormParams;
    try {
      const saveData = {
        form: currentFormId,
        user: userId,
        name: values?.name || trans.untitled,
        submitted: 0,
        duration: surveyDuration,
        json: values?.answers || {},
      };
      const dbCall = isNewSubmission
        ? crudDataPoints.saveDataPoint
        : crudDataPoints.updateDataPoint;
      const duration = getDurationInMinutes(surveyStart) + surveyDuration;
      await dbCall({
        ...currentDataPoint,
        ...saveData,
        duration: duration === 0 ? 1 : duration,
      });
      if (Platform.OS === 'android') {
        ToastAndroid.show(trans.successSaveDatapoint, ToastAndroid.LONG);
      }
      refreshForm();
      navigation.navigate('Home', { ...route?.params });
    } catch (err) {
      console.error(err);
      if (Platform.OS === 'android') {
        ToastAndroid.show(trans.errorSaveDatapoint, ToastAndroid.LONG);
      }
    }
  };

  const handleShowExitConfirmationDialog = () => {
    setShowDropdownMenu(false);
    setShowDialogMenu(false);
    setShowExitConfirmationDialog(true);
  };

  const handleOnExit = () => {
    refreshForm();
    return navigation.navigate('Home');
  };

  const handleOnSubmitForm = async (values) => {
    try {
      const answers = {};
      formJSON.question_group
        .flatMap((qg) => qg.question)
        .forEach((q) => {
          let val = values.answers?.[q.id];
          if (!val && val !== 0) {
            return;
          }
          if (q.type === 'cascade') {
            val = val.slice(-1)[0];
          }
          if (q.type === 'number') {
            val = parseFloat(val);
          }
          answers[q.id] = val;
        });
      // TODO:: submittedAt still null
      const submitData = {
        form: currentFormId,
        user: userId,
        name: values?.name || trans.untitled,
        geo: values.geo,
        submitted: 1,
        duration: surveyDuration,
        json: answers,
      };
      const dbCall = isNewSubmission
        ? crudDataPoints.saveDataPoint
        : crudDataPoints.updateDataPoint;
      const duration = getDurationInMinutes(surveyStart) + surveyDuration;
      await dbCall({
        ...currentDataPoint,
        ...submitData,
        duration: duration === 0 ? 1 : duration,
      });
      if (Platform.OS === 'android') {
        ToastAndroid.show(trans.successSubmitted, ToastAndroid.LONG);
      }
      refreshForm();
      navigation.navigate('Home', { ...route?.params });
    } catch (err) {
      console.error(err);
      if (Platform.OS === 'android') {
        ToastAndroid.show(trans.errorSubmitted, ToastAndroid.LONG);
      }
    }
  };

  return (
    <BaseLayout
      title={route?.params?.name}
      subTitle="formPage"
      leftComponent={
        <Button type="clear" onPress={handleOnPressArrowBackButton} testID="arrow-back-button">
          <Icon name="arrow-back" size={18} />
        </Button>
      }
      rightComponent={
        <SaveDropdownMenu
          visible={showDropdownMenu}
          setVisible={setShowDropdownMenu}
          anchor={
            <Button
              type="clear"
              testID="form-page-kebab-menu"
              onPress={() => setShowDropdownMenu(true)}
            >
              <Icon name="ellipsis-vertical" size={18} />
            </Button>
          }
          handleOnExit={handleShowExitConfirmationDialog}
          handleOnSaveAndExit={handleOnSaveAndExit}
        />
      }
    >
      {!loading ? (
        <FormContainer
          forms={formJSON}
          initialValues={currentValues}
          onSubmit={handleOnSubmitForm}
          onSave={onSaveCallback}
          setShowDialogMenu={setShowDialogMenu}
        />
      ) : (
        <View style={styles.loadingContainer}>
          <ActivityIndicator />
        </View>
      )}
      <SaveDialogMenu
        visible={showDialogMenu}
        setVisible={setShowDialogMenu}
        handleOnExit={handleShowExitConfirmationDialog}
        handleOnSaveAndExit={handleOnSaveAndExit}
      />
      <Dialog visible={showExitConfirmationDialog} testID="exit-confirmation-dialog">
        <Text testID="exit-confirmation-text">{trans.confirmExit}</Text>
        <Dialog.Actions>
          <Dialog.Button
            title={trans.buttonExit}
            onPress={handleOnExit}
            testID="exit-confirmation-ok"
          />
          <Dialog.Button
            title={trans.buttonCancel}
            onPress={() => setShowExitConfirmationDialog(false)}
            testID="exit-confirmation-cancel"
          />
        </Dialog.Actions>
      </Dialog>
    </BaseLayout>
  );
};

const styles = StyleSheet.create({
  loadingContainer: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
  },
});

export default FormPage;
