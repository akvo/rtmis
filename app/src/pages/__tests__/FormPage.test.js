import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitFor, act } from '@testing-library/react-native';
jest.useFakeTimers();
import FormPage from '../FormPage';
import { FormState } from 'store';

const mockFormContainer = jest.fn();
const mockRoute = {
  params: { id: 1, name: 'Form Name', newSubmission: true },
};
const mockNavigation = {
  navigate: jest.fn(),
  canGoBack: jest.fn(() => Promise.resolve(true)),
  goBack: jest.fn(),
};
const mockValues = {
  name: 'John',
  geo: null,
  answers: {
    1: 'John',
    2: new Date('01-01-1992'),
    3: '31',
    4: ['Male'],
    5: ['Bachelor'],
    6: ['Traveling'],
    7: ['Fried Rice'],
  },
};
const mockRefreshForm = jest.fn();

const exampleTestForm = {
  name: 'Testing Form',
  languages: ['en', 'id'],
  defaultLanguage: 'en',
  translations: [
    {
      name: 'Formulir untuk Testing',
      language: 'id',
    },
  ],
  question_group: [
    {
      name: 'Registration',
      order: 1,
      translations: [
        {
          name: 'Registrasi',
          language: 'id',
        },
      ],
      question: [
        {
          id: 1,
          name: 'Your Name',
          order: 1,
          type: 'input',
          required: true,
          meta: true,
          translations: [
            {
              name: 'Nama Anda',
              language: 'id',
            },
          ],
          addonBefore: 'Name',
        },
        {
          id: 2,
          name: 'Birth Date',
          order: 2,
          type: 'date',
          required: true,
          translations: [
            {
              name: 'Tanggal Lahir',
              language: 'id',
            },
          ],
        },
        {
          id: 3,
          name: 'Age',
          order: 3,
          type: 'number',
          required: true,
          translations: [
            {
              name: 'Umur',
              language: 'id',
            },
          ],
        },
        {
          id: 4,
          name: 'Gender',
          order: 4,
          type: 'option',
          required: true,
          option: [
            {
              id: 1,
              name: 'Male',
              order: 1,
            },
            {
              id: 2,
              name: 'Female',
              order: 2,
            },
          ],
          meta: false,
          translations: [
            {
              name: 'Jenis Kelamin',
              language: 'id',
            },
          ],
        },
        {
          id: 5,
          name: 'Your last education',
          order: 1,
          type: 'option',
          required: false,
          option: [
            {
              id: 11,
              name: 'Senior High School',
              order: 1,
            },
            {
              id: 12,
              name: 'Bachelor',
              order: 2,
            },
            {
              id: 13,
              name: 'Master',
              order: 3,
            },
            {
              id: 14,
              name: 'Doctor',
              order: 4,
            },
          ],
        },
        {
          id: 6,
          name: 'Hobby',
          order: 2,
          type: 'option',
          required: false,
          option: [
            {
              id: 21,
              name: 'Reading',
              order: 1,
            },
            {
              id: 22,
              name: 'Traveling',
              order: 2,
            },
            {
              id: 23,
              name: 'Programming',
              order: 3,
            },
          ],
        },
        {
          id: 7,
          name: 'Foods',
          order: 3,
          type: 'option',
          required: false,
          option: [
            {
              id: 31,
              name: 'Fried Rice',
              order: 1,
            },
            {
              id: 32,
              name: 'Rendang',
              order: 2,
            },
            {
              id: 33,
              name: 'Noodle',
              order: 3,
            },
            {
              id: 34,
              name: 'Meat Ball',
              order: 5,
            },
            {
              id: 35,
              name: 'Fried Chicken',
              order: 6,
            },
          ],
        },
        {
          id: 8,
          name: 'Comment',
          order: 4,
          type: 'text',
          required: false,
          translations: [
            {
              name: 'Kommentar',
              language: 'id',
            },
          ],
        },
        {
          id: 9,
          name: 'Give Rating from 1 - 9 for Rendang',
          order: 5,
          type: 'number',
          required: true,
          dependency: [
            {
              id: 7,
              options: ['Rendang'],
            },
          ],
          rule: {
            min: 1,
            max: 9,
            allowDecimal: true,
          },
          addonAfter: 'Score',
        },
      ],
    },
  ],
};

jest.mock('../../form/FormContainer', () => ({ forms, initialValues, onSubmit }) => {
  mockFormContainer(forms, initialValues, onSubmit);
  return (
    <mock-FormContainer>
      <button onPress={() => onSubmit(mockValues, mockRefreshForm)} testID="mock-submit-button">
        Submit
      </button>
    </mock-FormContainer>
  );
});

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useMemo: jest.fn(),
}));

describe('FormPage component', () => {
  test('should render component correctly', async () => {
    const tree = render(<FormPage navigation={mockNavigation} />);
    await waitFor(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('should render the FormPage with the correct form title', () => {
    const wrapper = render(<FormPage navigation={mockNavigation} route={mockRoute} />);
    expect(wrapper.getByText('Form Name')).toBeDefined();
  });

  test('should show the correct form content based on formJSON', async () => {
    FormState.useState.mockReturnValue({
      form: exampleTestForm,
    });
    jest.spyOn(React, 'useMemo').mockReturnValue(exampleTestForm);

    const wrapper = render(<FormPage navigation={mockNavigation} route={mockRoute} />);
    const { form: mockStateForm } = FormState.useState((s) => s);

    await waitFor(() => {
      expect(mockFormContainer.mock.calls[0]).toEqual([exampleTestForm, {}, expect.any(Function)]);
      expect(mockStateForm).toEqual(exampleTestForm);
      expect(wrapper.getByText('Form Name')).toBeDefined();
    });
  });
});
