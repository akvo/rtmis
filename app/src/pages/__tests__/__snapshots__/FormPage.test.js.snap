// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPage component should render component correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <ActivityIndicator />
  </View>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="save-dialog-menu"
    transparent={true}
    visible={false}
  />
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    style={
      {
        "width": "100%",
      }
    }
    testID="exit-confirmation-dialog"
    transparent={true}
    visible={false}
  />
</View>
`;
